### PR TITLE
 mds: add MAY_SNAPSHOT in MDSAuthCaps

### PR DESCRIPTION
**Pull request by ukernel**
For controlling whether a client is allowed to create or delete
snapshots

Fixes: http://tracker.ceph.com/issues/24284

### CODE REVIEW

File: PendingReleaseNotes [hunk 1]
```diff
@@ -18,3 +18,6 @@
 * The JSON output of the ``osd find`` command has replaced the ``ip``
   field with an ``addrs`` section to reflect that OSDs may bind to
   multiple addresses.
+
+* CephFS clients without the 's' flag in their authentication capability
+  string will no longer be able to create/delete snapshots.
```
Comment (batrick) [inline PendingReleaseNotes]: Let's add to this note what command needs to be run to enable snapshots for an existing credential.

**Comment by batrick** [comment]
````
2018-06-18 14:05:06.857 7f5e8d222700 15 mds.0.cache.ino(0x10000000002) project_inode 0x10000000002
2018-06-18 14:05:06.857 7f5e8d222700 10 mds.0.cache.dir(0x10000000001) pre_dirty 26
2018-06-18 14:05:06.857 7f5e8d222700 10 mds.0.cache.den(0x10000000001 1) pre_dirty [dentry #0x1/client.0/tmp/1 [2,head] auth (dversion lock) pv=26 v=22 ap=0+1 ino=0x10000000002 state=1610612736 | inodepin=1 dirty=1 0x56407ec78000]
2018-06-18 14:05:06.857 7f5e88a19700  1 -- 172.21.15.24:6805/1195466135 --> 172.21.15.24:6806/16814 -- osd_op(unknown.0.14:40 2.4 2:2e2fa760:::200.00000001:head [write 61058~1828 [fadvise_dontneed]] snapc 0=[] ondisk+write+known_if_redirected+full_force e28) v8 -- 0x56407ec18680 con 0
2018-06-18 14:05:06.857 7f5e8d222700 10 mds.0.cache.ino(0x10000000002) pre_dirty 26 (current v 22)
2018-06-18 14:05:06.857 7f5e8d222700 10 mds.0.cache.ino(0x10000000002) finish_scatter_gather_update 1024 on [inode 0x10000000002 [...6,head] /client.0/tmp/1/ auth v22 pv26 ap=1+0 snaprealm=0x56407eb28d00 dirtyparent f(v0 m2018-06-18 14:02:51.703345 2=1+1)->f(v0 m2018-06-18 14:02:51.703345 2=1+1) n(v1 rc2018-06-18 14:02:51.776661 4=1+3)->n(v1 rc2018-06-18 14:02:51.776661 4=1+3) (inest lock->sync w=1 dirty) (ifile excl) (iversion lock) caps={4275=pAsLsXsFsx/-@0},l=4275 | dirtyscattered=1 lock=1 dirfrag=1 caps=1 openingsnapparents=0 dirtyparent=1 dirty=1 authpin=1 0x56407ec6c700]
2018-06-18 14:05:06.857 7f5e8d222700 20 mds.0.cache.ino(0x10000000002)   orig rstat n(v1 rc2018-06-18 14:02:51.776661 4=1+3)
2018-06-18 14:05:06.857 7f5e8d222700 20 mds.0.cache.ino(0x10000000002) * [dir 0x10000000002 /client.0/tmp/1/ [2,head] auth v=17 cv=0/0 state=1610874881|complete f(v0 m2018-06-18 14:02:51.703345 2=1+1) n(v1 rc2018-06-18 14:02:51.776661 3=1+2) hs=2+0,ss=1+0 dirty=3 | child=1 dirty=1 0x56407ec61400]
2018-06-18 14:05:06.857 7f5e8d222700 10 mds.0.cache.dir(0x10000000002) project_fnode 0x56407e985e10
2018-06-18 14:05:06.857 7f5e8d222700 10 mds.0.cache.dir(0x10000000002) assimilate_dirty_rstat_inodes
2018-06-18 14:05:06.857 7f5e8d222700 10 mds.0.cache.dir(0x10000000002) assimilate_dirty_rstat_inodes done
2018-06-18 14:05:06.857 7f5e8d222700 20 mds.0.cache.ino(0x10000000002) *           rstat n(v1 rc2018-06-18 14:02:51.776661 3=1+2)
2018-06-18 14:05:06.857 7f5e8d222700 20 mds.0.cache.ino(0x10000000002) * accounted_rstat n(v1 rc2018-06-18 14:02:51.776661 3=1+2)
2018-06-18 14:05:06.857 7f5e8d222700 20 mds.0.cache.ino(0x10000000002) * dirty_old_rstat {}
2018-06-18 14:05:06.857 7f5e8d222700 10 mds.0.cache project_rstat_frag_to_inode [2,head]
2018-06-18 14:05:06.857 7f5e8d222700 20 mds.0.cache   frag           rstat n(v1 rc2018-06-18 14:02:51.776661 3=1+2)
2018-06-18 14:05:06.857 7f5e8d222700 20 mds.0.cache   frag accounted_rstat n(v1 rc2018-06-18 14:02:51.776661 3=1+2)
2018-06-18 14:05:06.857 7f5e8d222700 20 mds.0.cache                  delta n(v1 rc2018-06-18 14:02:51.776661)
2018-06-18 14:05:06.857 7f5e8d222700 20 mds.0.cache  projecting to [6,head] n(v2 rc2018-06-18 14:02:51.776661 4=1+3)
2018-06-18 14:05:06.857 7f5e8d222700 20 mds.0.cache         result [6,head] n(v2 rc2018-06-18 14:02:51.776661 4=1+3)
2018-06-18 14:05:06.857 7f5e8d222700 20 mds.0.cache  projecting to [5,5] n(v1 rc2018-06-18 14:02:51.776661 4=1+3)
2018-06-18 14:05:06.857 7f5e8d222700 20 mds.0.cache         result [5,5] n(v1 rc2018-06-18 14:02:51.776661 4=1+3)
2018-06-18 14:05:06.857 7f5e8d222700 20 mds.0.cache  projecting to [3,4] n(v0 rc2018-06-18 14:02:51.776661 4=1+3)
2018-06-18 14:05:06.857 7f5e8d222700 20 mds.0.cache         result [3,4] n(v0 rc2018-06-18 14:02:51.776661 4=1+3)
2018-06-18 14:05:06.857 7f5e8d222700 20 mds.0.cache  projecting to [2,2] n(v0 rc2018-06-18 14:02:51.776661 4=1+3)
2018-06-18 14:05:06.857 7f5e8d222700 20 mds.0.cache         result [2,2] n(v0 rc2018-06-18 14:02:51.776661 4=1+3)
2018-06-18 14:05:06.857 7f5e8d222700 20 mds.0.cache.dir(0x10000000002) get_num_head_items() = 2; fnode.fragstat.nfiles=1 fnode.fragstat.nsubdirs=1
2018-06-18 14:05:06.857 7f5e8d222700 20 mds.0.cache.dir(0x10000000002) total of child dentrys: n(v0 rc2018-06-18 14:02:51.776661 3=1+2)
2018-06-18 14:05:06.857 7f5e8d222700 20 mds.0.cache.dir(0x10000000002) my rstats:              n(v1 rc2018-06-18 14:02:51.776661 3=1+2)
2018-06-18 14:05:06.857 7f5e8d222700 10 mds.0.cache.dir(0x10000000002) check_rstats complete on 0x56407ec61400
2018-06-18 14:05:06.857 7f5e8d222700 20 mds.0.cache.dir(0x10000000002) get_num_head_items() = 2; fnode.fragstat.nfiles=1 fnode.fragstat.nsubdirs=1
2018-06-18 14:05:06.857 7f5e8d222700 20 mds.0.cache.dir(0x10000000002) total of child dentrys: n(v0 rc2018-06-18 14:02:51.776661 3=1+2)
2018-06-18 14:05:06.857 7f5e8d222700 20 mds.0.cache.dir(0x10000000002) my rstats:              n(v1 rc2018-06-18 14:02:51.776661 3=1+2)
2018-06-18 14:05:06.857 7f5e8d222700 10 mds.0.cache.dir(0x10000000002) check_rstats complete on 0x56407ec61400
2018-06-18 14:05:06.857 7f5e8d222700 10 mds.0.cache.ino(0x10000000002) * updated accounted_rstat n(v2 rc2018-06-18 14:02:51.776661 3=1+2) on [dir 0x10000000002 /client.0/tmp/1/ [2,head] auth pv=17 v=17 cv=0/0 state=1610907649|complete|assimrstat f(v0 m2018-06-18 14:02:51.703345 2=1+1)->f(v0 m2018-06-18 14:02:51.703345 2=1+1) n(v1 rc2018-06-18 14:02:51.776661 3=1+2)->n(v2 rc2018-06-18 14:02:51.776661 3=1+2) hs=2+0,ss=1+0 dirty=3 | child=1 dirty=1 0x56407ec61400]
2018-06-18 14:05:06.857 7f5e8d222700 20 mds.0.cache.ino(0x10000000002)  final rstat n(v2 rc2018-06-18 14:02:51.776661 4=1+3)
2018-06-18 14:05:06.857 7f5e8d222700 -1 log_channel(cluster) log [ERR] : inconsistent rstat on inode 0x10000000002, inode has n(v2 rc2018-06-18 14:02:51.776661 4=1+3), directory fragments have n(v0 rc2018-06-18 14:02:51.776661 rs2 4=1+3)
2018-06-18 14:05:06.861 7f5e92a2d700  1 -- 172.21.15.24:6805/1195466135 <== osd.4 172.21.15.24:6807/16826 4 ==== osd_op_reply(37 mds0_openfiles.0 [omap-clear,omap-set-header,omap-set-vals] v28'7 uv7 ondisk = 0) v8 ==== 244+0+0 (3277418390 0 0) 0x56407ec1b740 con 0x56407ec2c700
2018-06-18 14:05:06.861 7f5e89a1b700 10 MDSIOContextBase::complete: 16C_IO_OFT_Journal
2018-06-18 14:05:06.861 7f5e91a2b700  1 -- 172.21.15.24:6805/1195466135 <== osd.5 172.21.15.24:6806/16814 10 ==== osd_op_reply(38 200.00000001 [write 57674~1551 [fadvise_dontneed]] v28'43 uv43 ondisk = 0) v8 ==== 156+0+0 (2891576637 0 0) 0x56407ec18680 con 0x56407ec2dc00
2018-06-18 14:05:06.861 7f5e8d222700 -1 /build/ceph-14.0.0-583-g44db180/src/mds/CInode.cc: In function 'void CInode::finish_scatter_gather_update(int)' thread 7f5e8d222700 time 2018-06-18 14:05:06.863034
/build/ceph-14.0.0-583-g44db180/src/mds/CInode.cc: 2346: FAILED assert(!"unmatched rstat" == g_conf->mds_verify_scatter)

 ceph version 14.0.0-583-g44db180 (44db1805f8b3d0f2ca39b5967dd715b93ad26b81) nautilus (dev)
 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x102) [0x7f5e976ce6b2]
 2: (()+0x2e6877) [0x7f5e976ce877]
 3: (CInode::finish_scatter_gather_update(int)+0x29eb) [0x56407cae8a2b]
 4: (Locker::scatter_writebehind(ScatterLock*)+0x13b) [0x56407ca3e1db]
 5: (Locker::simple_sync(SimpleLock*, bool*)+0x176) [0x56407ca423c6]
 6: (Locker::scatter_nudge(ScatterLock*, MDSInternalContextBase*, bool)+0x3dd) [0x56407ca46e6d]
 7: (Locker::scatter_tick()+0x1e4) [0x56407ca474a4]
 8: (Locker::tick()+0x9) [0x56407ca477b9]
 9: (MDSRankDispatcher::tick()+0x1d4) [0x56407c8e5914]
 10: (FunctionContext::finish(int)+0x2c) [0x56407c8d5b5c]
 11: (Context::complete(int)+0x9) [0x56407c8d3bc9]
 12: (SafeTimer::timer_thread()+0x18b) [0x7f5e976cb02b]
 13: (SafeTimerThread::entry()+0xd) [0x7f5e976cc5ed]
 14: (()+0x76ba) [0x7f5e96f616ba]
 15: (clone()+0x6d) [0x7f5e9678a41d]
 NOTE: a copy of the executable, or `objdump -rdS <executable>` is needed to interpret this.
````

/ceph/teuthology-archive/pdonnell-2018-06-18_13:12:11-fs-wip-pdonnell-testing-20180615.150516-testing-basic-smithi/2677572/remote/smithi024/log/ceph-mds.b.log.gz

**Comment by ukernel** [comment]
the crash (incorrect rsnaps during upgrade test) is expected when mds_verify_scatter is true. when mds_verify_scatter is false (default), mds will fix incorrect rsnaps

**Comment by batrick** [comment]
> the crash (incorrect rsnaps during upgrade test) is expected when mds_verify_scatter is true.

I don't see `mds_verify_scatter` being turned on in that test. Can you point to where? In any case, I was going to suggest turning it off until the MDS has a chance to fix the incorrect rsnaps.

**Comment by batrick** [comment]
@ukernel PTAL

**Comment by ukernel** [comment]
it's true for teuthology job, see teuthology/teuthology/ceph.conf.template

**Comment by ukernel** [comment]
@batrick  updated

**Comment by batrick** [comment]
http://pulpito.ceph.com/pdonnell-2018-06-27_19:41:49-fs:upgrade-wip-pdonnell-testing-20180627.172959-testing-basic-smithi/

**Pull request merged**